[DOCS-13682] Expand Understanding Injector Behavior with Windows, life cycle, and privilege details (#36552)
DataDog/documentation: content/en/tracing/guide/injectors.md

The default delivery method uses init containers, but SSI supports additional [injection modes][2] that may better suit your environment.

### Windows

> On Windows, the injector is delivered as a DLL. A Datadog kernel driver loads the injector DLL into every new process, and the DLL's `DllMain` function runs before the application's entry point. The injector then performs the same detection and instrumentation steps as on Linux.

Reviewer: >The injector then performs the same detection and instrumentation steps as on Linux.

We don't actually explain these under Linux (unless I missed it)
But also the detection mechanism on Windows is a bit different, the injection is for the most part the same.

Author: Reworded to drop the "same detection as Linux" claim, point readers to the Injector behavior section, note that Windows supports only Java and .NET, and acknowledge detection differs in implementation. In 898340b9db.

## Injector behavior

After the injector is loaded into a process's memory space—regardless of platform—it performs the following:
- Detects the application's runtime (such as Python, Node.js, PHP, Java, .NET, Ruby, or NGINX).

> - Evaluates whether the process should be instrumented.

Reviewer: should we have links here to a/the doc explaining what we exclude and how to use [instrumentation rules](https://docs.datadoghq.com/tracing/trace_collection/single-step-apm/linux/#define-instrumentation-rules) ?

Author: Added a sentence noting default exclusions plus links to the instrumentation-rules sections on Linux, Kubernetes, and Windows. In 898340b9db.

- Modifies environment variables or command-line arguments to load the appropriate tracer SDK.
- Emits telemetry to report status and aid in debugging.

If injection fails or no supported runtime is detected, the application continues to run without instrumentation. The same applies if a process is excluded by configuration.

**Note**: Because the injector runs before the SDK evaluates its environment variables, setting `DD_TRACE_ENABLED=false` does not prevent SSI from loading the SDK. To disable or remove SSI, see the [Single Step APM Instrumentation][1] setup page for your platform.

## Injector life cycle

> The injector runs before the application's `main` function and exits before the application starts. It is not a background process or thread that runs alongside the application.

Reviewer: ```suggestion
The injector completes its work before the application's main function runs. It is not a background process or thread that runs alongside the application.
```

it might just be me, "exits" sounds to me like it also unloads or something, it confuses me (to be clear it is still mapped to memory during the whole program life cycle)

Author: Applied — switched "exits before the application starts" to "completes its work before the application's `main` function runs." In 898340b9db.

> After the injector sets the necessary environment variables (or, for NGINX, re-executes the process with the Datadog module loaded), the injected library completes its work and the application proceeds normally. From that point on, the only Datadog code in the process is the tracer SDK that the application itself loads through the configured environment variables.

Reviewer: ```suggestion
After the injector sets the necessary environment variables/command line arguments, the injected library completes its work and the application proceeds normally. From that point on, the only Datadog code in the process is the tracer SDK that the application itself loads through the configured environment variables.
```

Author: Applied with light edits: dropped the NGINX parenthetical and folded into "environment variables or command-line arguments," and softened the second sentence to "using that configuration" since for NGINX the tracer loads through a command-line argument rather than an env var. In 898340b9db.

## Per-language instrumentation

To load the appropriate tracer SDK, the injector sets one or more environment variables for the application process. The mechanism varies by runtime:

| Runtime  | Mechanism |
|----------|-----------|
| Java     | Sets `JAVA_TOOL_OPTIONS` with `-javaagent:<path-to-dd-java-agent.jar>`. |
| Python   | Sets `PYTHONPATH` to the tracer installation directory. |
| Node.js  | Sets `NODE_OPTIONS` with `--require` (CommonJS) or `--import` (ECMAScript modules). |
| Ruby     | Sets `RUBYOPT` with `-r <path-to-Datadog-host-inject-script>`. |
| .NET     | Sets the .NET CLR profiler environment variables (`CORECLR_ENABLE_PROFILING`, `CORECLR_PROFILER`, `CORECLR_PROFILER_PATH`, and related variables). |
| PHP      | Sets `PHP_INI_SCAN_DIR` to a Datadog-provided loader directory. |

> | NGINX    | Re-executes the NGINX binary with the Datadog module loaded (for example, `-g "load_module <path-to-Datadog-module>;"`). Because NGINX is a web server that loads native modules rather than a language runtime, it cannot be instrumented with environment variables. |

Reviewer: ```suggestion
| NGINX    | Modifies the command line arguments with the Datadog module loaded (for example, `-g "load_module <path-to-Datadog-module>;"`).|
```

Author: Kept the re-exec verb (the process replacement seemed worth preserving) but folded in the command-line-argument framing and dropped the explanatory clause. In 898340b9db.

## Privileges and scope

### Injection scope

> SSI instruments applications by loading a tracer or profiler into the application process using user-space dynamic linking (such as `LD_PRELOAD` on Linux or DLL injection on Windows). It does not install kernel modules, patch the kernel, hook system calls, or use eBPF.

Reviewer: it might be worth mentioning somewhere\somehow that the injector inherits the process privileges it is injected into

Author: Added a sentence at the top of Injection scope surfacing that the injector, tracer, and profiler all inherit the application process's OS privileges, with a forward link to Runtime privileges. In 898340b9db.

While the injected code runs inside the application process, some deployment modes update OS or runtime configuration so that the injector loads automatically:

> - **Kubernetes admission controller injection**: Mutates pod specs (adds an init container, mounts volumes, sets environment variables). This is a Kubernetes API action, not a host OS modification.

Reviewer: injecting init containers is one way, the CSI driver is another and we're leaning towards making that the default way.

@iamluc what do you think?

Author: Broadened the bullet to mention both init containers and the Datadog CSI driver (linking to the existing injection-modes reference). Stayed neutral on which is the default since your note suggested CSI is becoming default but isn't there yet — happy to revisit once @iamluc weighs in. In 898340b9db.

> - **Linux host injection**: Configures the dynamic loader (typically through `/etc/ld.so.preload`) so that new processes load the injection launcher. This is an OS-level configuration change, but it remains user-space dynamic linking, not kernel code.

Reviewer: ```suggestion
- **Linux host injection**: Configures the dynamic loader (typically through `/etc/ld.so.preload`) so that every new dynamically linked processes load the injection launcher. This is an OS-level configuration change, but it remains user-space dynamic linking, not kernel code.
```

Author: Applied with a small grammar tweak ("every new dynamically linked process loads"). In 898340b9db.

> - **Docker runtime injection**: Configures Docker to use a wrapper runtime so that containers start with the tracer mounted and configured. This modifies Docker daemon configuration. It does not modify kernel modules.

Reviewer: ```suggestion
- **Docker runtime injection**: Configures Docker to use a wrapper runtime so that containers start with the tracer mounted and configured by injecting an `LD_PRELOAD` environment variable into the container. This modifies Docker daemon configuration.
```
I'm not sure I understand in what context we added kernel modules here

Author: Applied — added the `LD_PRELOAD` detail and dropped "It does not modify kernel modules" since the Injection scope intro already covers that for all deployment modes. In 898340b9db.

- **Windows injection**: Installs a kernel driver that loads the injector DLL into new processes. The driver runs at the OS level; the injected code that runs inside each application process remains user-space.

### Runtime privileges

> After injection, the tracer or profiler runs inside the application process and inherits the same OS permissions as that process. It does not gain additional privileges. If the application runs as a restricted user, the tracer or profiler is also restricted to that user's permissions.

Reviewer: Maybe we should start with the injector inherits the process privileges, as well as the tracers/profilers...

Author: Reordered Runtime privileges to lead with inheritance and broadened the subject to include the injector (not just tracer/profiler). The Injection scope intro now also leads with this point per your comment above. In 898340b9db.

### Installation privileges

Privilege requirements at install time depend on the deployment mode:

#### Kubernetes admission controller

> - **Cluster-side components** (the Datadog Cluster Agent and admission controller) typically require cluster-level permissions to register webhooks and mutate pods.

Reviewer: @iamluc

- **Application pods** do not require elevated OS privileges from Datadog, but hardened clusters may need to allow:
  - Volume mounts used to communicate with the Agent (commonly a `hostPath` mount to a socket directory).
  - Specific `SecurityContext` or SCC allowances.

  You can keep the injected init container and application container locked down (non-root, dropped capabilities, no privilege escalation) as long as your policies permit the required mounts.

#### Linux host

- Installation typically requires root (or equivalent), because it installs packages and may modify system-wide loader configuration and service configuration.
- After installation, the Datadog Agent service runs as its own service user (commonly `dd-agent`), and the injected code runs with the application user's privileges.

#### Docker runtime on a host

- Installation and configuration typically require root, because they modify Docker daemon configuration.

> - At runtime, injection affects container startup behavior. The tracer runs inside the container's application process as that container's user.

Reviewer: ```suggestion
- At runtime, injection affects container startup behavior. The injector and the tracer runs inside the container's application process as that container's user.
```

Author: Applied with grammar fix ("the injector and the tracer run"). In 898340b9db.
